packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4585,652 +4353,1109 @@ mixin RelayoutWhenSystemFontsChangeMixin on RenderObject {
+    final List<_SemanticsFragment> children = <_SemanticsFragment>[];
+    siblingMergeGroups.clear();
+    _childrenAndElevationAdjustments.clear();
+    mergeUp.clear();
+    for (final _RenderObjectSemantics childSemantics in _getNonBlockedChildren()) {
+      assert(!childSemantics.renderObject._needsLayout);
+      final _SemanticsParentData childParentData = _SemanticsParentData(
+        // Need to use original config as the effective config may have been
+        // modified.
+        mergeIntoParent: (parentData?.mergeIntoParent ?? false) || configProvider.original.isMergingSemanticsOfDescendants,
+        blocksUserActions: (parentData?.blocksUserActions ?? false) || configProvider.original.isBlockingUserActions,
+        explicitChildNodes: explicitChildNodesForChildren,
+        tagsForChildren: tagsForChildren,
+      );
+      childSemantics.didUpdateParentData(childParentData);
+      for (final _SemanticsFragment fragment in childSemantics.mergeUp) {
+        if (hasChildConfigurationsDelegate && fragment.configToMergeUp != null) {
+          // This fragment need to go through delegate to determine whether it
+          // merge up or not.
+          childConfigurations.add(fragment.configToMergeUp!);
+          configToFragment[fragment.configToMergeUp!] = fragment;
+        } else {
+          children.add(fragment);
+        }
+      }
 
-    if (!_mergeIntoParent && (geometry?.dropFromTree ?? false)) {
-      return; // Drop the node, it's not going to be visible.
+      if (!childSemantics.contributesToSemanticsTree) {
+        // This child semantics needs to propagate sibling merge group to be
+        // compiled by parent that contributes to semantics tree.
+        siblingMergeGroups.addAll(childSemantics.siblingMergeGroups);
+      }
     }
 
-    final SemanticsNode node = (owner._semantics ??= SemanticsNode(showOnScreen: owner.showOnScreen))
-      ..tags = _tagsForChildren;
+    assert(childConfigurationsDelegate != null || configToFragment.isEmpty);
+    if (!explicitChildNodesForChildren && hasChildConfigurationsDelegate) {
+      final ChildSemanticsConfigurationsResult result = childConfigurationsDelegate(childConfigurations);
+      children.addAll(
+        result.mergeUp.map<_SemanticsFragment>((SemanticsConfiguration config) {
+          final _SemanticsFragment? semantics = configToFragment[config];
+          if (semantics != null) {
+            return semantics;
+          }
+          return _IncompleteSemanticsFragment(config, this);
+        }),
+      );
+      for (final Iterable<SemanticsConfiguration> group in result.siblingMergeGroups) {
+        siblingMergeGroups.add(
+          group.map<_SemanticsFragment>((SemanticsConfiguration config) {
+            return configToFragment[config] ?? _IncompleteSemanticsFragment(config, this);
+          }).toList(),
+        );
+      }
+    }
+    mergeUp.addAll(children);
```

Comment:
It feels backwards (and probably extra work) to add items to the `mergeUp` list and later do:
```dart
mergeUp.clear();
mergeUp.add(this);
```
Why not something like this:
```dart
// mergeUp.addAll(children); // <-- remove this line
if (contributesToSemanticsTree) {
  // ...
  _marksExplicitInMergeGroup(children, isMergeUp: true);
  // ...
  mergeUp.add(this);
} else {
  mergeUp.addAll(children);
}
```

Reply:
mergeUp is used in the if case to get the mergeUpConfigs.